sdk/src/logs/batch_log_processor.cc

```diff
@@ -151,19 +153,28 @@ void BatchLogProcessor::Export(const bool was_force_flush_called)
                       return true;
                     });
                   });
+  if (is_export_async_ == false || was_force_flush_called == true) {
```

Comment:
I think async export should also be called in case of force flush if `is_export_async` is true. The callback should do the notify in that case?

Reply:
Agreed on force flush. What is your thoughts in case of Shutdown?
Since shutdown is called during destructor, do we need use async export during shutdown.

Reply:
No, there should be no new upload during the shutdown, it should just wait on the thread for pending uploads to finish.

Reply:
Shouldn't we wait util all data to be exported when we call ForceFlush or Shutdown when in async mode?

Reply:
Agreed @owent, we should wait.

Reply:
Agree we should wait in both cases. The [notification ](https://github.com/open-telemetry/opentelemetry-cpp/blob/49648bd8f741a28a2b111c8d8ecad75a070ea649/sdk/src/trace/batch_span_processor.cc#L163) logic can be called from callback for forceflush(), and similar logic can be added for shutdown() ?

Reply:
Yes, @lalitb in case of shutdown, we need to add  a new notification logic. I  will do that and raise a PR again

Reply:
> Yes, @lalitb in case of shutdown, we need to add a new notification logic. I will do that and raise a PR again

On further thought, this will work for the export triggered by the forceflush/shutdown. But how about the already ongoing async requests. We need to wait for their completion too. This means that the processor needs to keep track of the status of those requests.

Reply:
Thats a valid point @lalitb. What suggestions do you have on achieving this?  Should the processor create thread and call async export on the new thread?

Reply:
This probably needs more design thinking before finalizing. The callback for the status of export would be invoked by a thread outside of the processor, and this should somehow update the status to the processor. Will give more thoughts on this before writing :)

Reply:
> This probably needs more design thinking before finalizing. The callback for the status of export would be invoked by a thread outside of the processor, and this should somehow update the status to the processor. Will give more thoughts on this before writing :)

The same problem is already solved in `OtlpHttpClient::Export` of #1209 . When there are too many sessions, `OtlpHttpClient::Export` will wait for some some sessions to finish and then return.

As it in  https://github.com/open-telemetry/opentelemetry-cpp/pull/1209/files#diff-eb7f99ebc152d9cc0f4781cc0a0388907e4d4b58271dba85eba99fd16b880674R782 .The thread which call `OtlpHttpClient::Export` will use one lock and one condition variable to solve the concurrency problem. The lock is used to protect the management of running sessions, and the condition variable is used to be waked up when any session is finished.We allow any thread to call `Shutdown/ForceFlush` or finish and wake up the exporting thread. According to the spec. only one thread will call `Export`.

I think we can do the same changes here.

**BTW:** Can we create a feature branch for #1209 and this PR and merge #1209 to this new branch? We both change  codes for concurrency IO. There will be more and more conflicts if we do more changes later.The later we finish these two PRs, the harder to merge the conflicts.

Reply:
> The same problem is already solved in OtlpHttpClient::Export of https://github.com/open-telemetry/opentelemetry-cpp/pull/1209 . When there are too many sessions, OtlpHttpClient::Export will wait for some some sessions to finish and then return.

Thanks, if this ensures that only one thread will call Export() and works well for new export interface :) Good idea to have both the changes in the feature branch.

Reply:
Sure, I would look up the changes in OtlpHttpClient::Export  and would bring in those concept to Processor. 
Let me know If the feature branch is created and step forward to merge them.

Reply:
I have looked into the changes done in #1209. Since OtlpHttpClient had to maintain Sessions and EventHandler object until the export completes, it had achieved this by having map of running sessions and list of completed sessions to deal with it and its associated logic to clean up. Thanks @owent for sharing this PR.

From the Processor perspective, we do not need any object to keep it alive till the completion of async Export. All it needs is to wait for uncompleted async Exports to complete during shutdown and we can know this only by the callback. To achieve this, we can add an atomic counter -- which can be increased on every async Export call and would be decreased on the callback function. And Shutdown would wait until this counter is zero. 
        Counter = 0   ---> means no async export running and shutdown can proceed forward
        Counter = n, (n > 0). ----> means n no of async export/thread running concurrently.
        
 @owent @lalitb Please provide your arguments on this.

Reply:
I suggest to maintain the exporting recordable objects in process.When any exporter have BUGs which try to finish some futures more than once or do not finish some futures for ever, the Shutdown/ForceFlush may wait for ever.

Please let me known if you have any thought about this.

Reply:
Agreed @owent. We should maintain the recordable object and also not wait forever during Shutdown/ForceFlush. Thanks.
What is the plan to merge both PRs into a feature branch?

Reply:
> What is the plan to merge both PRs into a feature branch?

We can create a feature branch on the repo ( not on its fork), and you can keep merging the PRs on that. Finally, we can merge that feature to the main. I don't think creating branch on main repo is restricted, otherwise let me know I can create the branch @owent @DebajitDas

Reply:
@lalitb I think I do not have rights to create a branch. Can you please help creating a branch?

Reply:
@owent @DebajitDas  have created a brach `async-changes` in the main repo. You can create the PR on this branch ( instead of main) to push your changes there. Anyone should be able to merge to this branch once PR is created. PR to merge from `async-changes` to `main` branch can be created once the changes are ready.

Reply:
I have create #1274 to merge "Cocurrency otlp http session" into `[async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes)` .

Reply:
Hi @owent, I have raised https://github.com/open-telemetry/opentelemetry-cpp/pull/1275, but how do we want to proceed? Shall I go merge my changes first?

Reply:
@lalitb Could you please review #1274 or #1275 and merge any of them into [async-changes](https://github.com/open-telemetry/opentelemetry-cpp/tree/async-changes) first. And then we can continue to rebase the other one.
@DebajitDas I think it's fine to merge any of these two PRs first.

Reply:
Sure, have merged #1274. Looks like need to fix the conflict for #1275 to merge that.

Reply:
Resolved conflict on #1275. Please let me know if this can be merged. @lalitb